jax/interpreters/pxla.py

```diff
@@ -823,12 +859,21 @@ def get_num_partitions(*partitions):
 class ResultToPopulate(object): pass
 result_to_populate = ResultToPopulate()
 
-def _pvals_to_results_handler(size, nrep, out_pvals, devices, backend):
+def _pvals_to_results_handler(
+    size, nrep, npart,
+    out_parts: Optional[Tuple[PartitionsOrReplicated, ...]],
+    out_pvals, devices, backend):
   nouts = len(out_pvals)
-  handlers = [_pval_to_result_handler(size, nrep, pval, devices, backend)
-              for pval in out_pvals]
+  if out_parts is None:
+    out_parts = (None,) * len(out_pvals)
+  handlers = [
+      _pval_to_result_handler(size, nrep, npart, parts, pval, devices, backend)
+      for pval, parts in safe_zip(out_pvals, out_parts)
+  ]
+
   def handler(out_bufs):
-    buffers = [[result_to_populate] * nrep for _ in range(nouts)]
+    assert nrep * npart == len(out_bufs)
+    buffers = [[result_to_populate] * nrep * npart for _ in range(nouts)]
```

Comment:
Isn't it possible to construct the buffers directly from out_bufs and tuple_buf, without allocating the lists and then updating them? Then we can just use tuples which are faster that lists.

Reply:
Perhaps! I have a microbenchmark for this path in pmap_benchmark.py, we could try it out. I'm gonna leave the logic as-is for this change though.